desktop_version/src/DiscordNetwork.c

```diff
@@ -0,0 +1,165 @@
+} app;
+
+struct DiscordActivity activity;
+
+int discordCrashes = 0; // This is here to ensure we do not get stuck in a theoratical softlock of opening and crashing Discord instances.
+
+static void* libHandle = NULL;
+
+
+#define FUNC_LIST \
+    FOREACH_FUNC(enum EDiscordResult, DiscordCreate, (DiscordVersion, struct DiscordCreateParams*, struct IDiscordCore**))
+
+
+
+#define FOREACH_FUNC(rettype, name, params) static rettype (*name) params = NULL;
+FUNC_LIST
+#undef FOREACH_FUNC
+
+static void ClearPointers(void)
+{
+    SDL_UnloadObject(libHandle);
+    libHandle = NULL;
+#define FOREACH_FUNC(rettype, name, params) name = NULL;
+    FUNC_LIST
+#undef FOREACH_FUNC
+}
+
+void DISCORD_shutdown(void)
+{
+    if (libHandle)
+    {
+        ClearPointers();
+    }
+}
+
+int32_t DISCORD_init(void)
+{
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__HAIKU__) || defined(__DragonFly__)
+    return 0;
+#endif
+    libHandle = SDL_LoadObject(DISCORD_LIBRARY);
+    if (!libHandle)
+    {
+        vlog_info(DISCORD_LIBRARY " not found!");
+        printf("Can't load object %s : %s\n", DISCORD_LIBRARY, SDL_GetError());
+        return 0;
+    }
+
+#define FOREACH_FUNC(rettype, name, params) \
+    name = (rettype (*) params) (intptr_t) SDL_LoadFunction(libHandle, #name); \
+    if (!name) \
+    { \
+        vlog_error(DISCORD_LIBRARY " symbol " #name " not found!"); \
+        ClearPointers(); \
+        return 0; \
+    }
+    FUNC_LIST
+#undef FOREACH_FUNC
+    memset(&app, 0, sizeof(app));
+
+    struct DiscordCreateParams params;
+    params.client_id = DISCORD_CLIENT_ID;
+    params.flags = DiscordCreateFlags_Default;
+
+    if(!DiscordCreate(DISCORD_VERSION, &params, &app.core))
+    {
+        return 1;
+    }
+
+
+    strncpy(activity.state, "On the Main Menu", 128);
+    //strncpy(activity.details, "I'm Sorry", 128);
+    strncpy(activity.assets.large_image, "vvvvvv", 128);
+    strncpy(activity.assets.large_text, "Dimension  VVVVVV", 128);
+
+    app.activityMan = app.core->get_activity_manager(app.core);
+    app.activityMan->update_activity(app.activityMan, &activity, NULL, NULL);
+
+
+
+    return 0;
+}
+
+void DISCORD_REQUIRE(int x) {
+    if(discordCrashes > DISCORD_CRASH_LIMIT)
+    {
+        DISCORD_shutdown();
+        return;
+    }
+    if(x != DiscordResult_Ok)
+    {
+        ++discordCrashes;
+        DISCORD_shutdown();
+        DISCORD_init();
+        return;
+    }
+    discordCrashes = 0;
+}
+
+void DISCORD_update(const char *level, const char *name)
+{
+    if(discordCrashes > DISCORD_CRASH_LIMIT) {
+        return;
+    }
+    if(app.activityMan == NULL)
+    {
+        app.activityMan = app.core->get_activity_manager(app.core);
+    }
+    strncpy(activity.state, name, 127);
+    strncpy(activity.assets.large_image, "vvvvvv", 127);
+    strncpy(activity.assets.large_text, level, 127);
+    activity.state[127] = '\0';
+    activity.assets.large_image[127] = '\0';
+    activity.assets.large_text[127] = '\0';
```

Comment:
Instead of `strncpy` you can use `SDL_strlcpy` - which isn't so broken and doesn't require a separate `'\0'` assignment. You should also use `sizeof` instead of copying the buffer length. i.e.:

```suggestion
    SDL_strlcpy(activity.state, name, sizeof(activity.state));
    SDL_strlcpy(activity.assets.large_image, "vvvvvv", sizeof(activity.assets.large_image));
    SDL_strlcpy(activity.assets.large_text, level, sizeof(activity.assets.large_text));
```

(ditto the other place)